### PR TITLE
fix(cli): Add fallback method for determining internal IP

### LINKED ISSUE #24759: Unable to make expo listen to LAN port because WMIC is missing
### Summary

When running 
* `npx expo start`
* `npx expo`
* `npx expo --host lan`
* `npx expo --lan`

It always goes Metro waiting on exp://127.0.0.1:8081 and connecting using Expo Go fails because there's nothing listening on 127.0.0.1 on the device.

* `npx expo --tunnel` works

### What platform(s) does this occur on?

iOS

### SDK Version

49.0.13

### Environment

Bunch of errors but ended up narrowing to
>   "'wmic' is not recognized as an internal or external command,\r\n" +
>    'operable program or batch file.\r\n',

There's nothing in the [requirements](https://docs.expo.dev/get-started/installation/#requirements) that state that optional feature (it's optional in Windows 11) is required.

After installing and rerun I got this
```
 expo-env-info 1.0.5 environment info:
    System:
      OS: Windows 10 10.0.22631
    Binaries:
      Node: 18.18.0 - ~\.nvm\versions\node\v18.18.0\bin\node.EXE
      Yarn: 1.22.19 - ~\.nvm\versions\node\v18.18.0\bin\yarn.CMD
      npm: 9.8.1 - ~\.nvm\versions\node\v18.18.0\bin\npm.CMD
      Watchman: 20230806.083715.0 - C:\Users\trajano\scoop\shims\watchman.EXE
    IDEs:
      Android Studio: AI-223.8836.35.2231.10406996
    npmPackages:
      expo: ~49.0.13 => 49.0.13
      react: 18.2.0 => 18.2.0
      react-native: 0.72.5 => 0.72.5
    Expo Workflow: managed
```

### Minimal reproducible example

I just did 
```
nvm use --lts
npm i -g yarn
yarn create expo my-app
npx expo start
```

Then tried to go on Expo Go.

### PR DESCRIPTION
**Pull request by kitten**
# Why

The `default-gateway` package that `internal-ip` relies on is going out of maintenance and has rejected an issue to fix getting the gateway on Windows, which it usually does using the `wmic` command. The `default-gateway` package basically maintains a large set of commands it relies on to get the default gateway per platform.
However, Windows won't always come with the `wmic` command since Windows 10, and this method of determining the IP via the default gateway will hence fail.

Instead of relying only on `internal-ip` & `default-gateway`, we can implement an alternative approach to resolve the default IP.

Resolves #24759

# How

By opening a socket to a publicly routed IP address (i.e. WAN), we can determine the local IP address that's bound to that socket. (Basically, `getsockname` on Unix systems). This is guaranteed to use the system's default gateway.

To workaround the issue that opening a TCP socket incurs a potential delay while a connection is established, we may instead use a UDP socket, which shouldn't send any packets to open the socket.

Lastly, we're wrapping this in a Node child process via `spawnSync` to maintain the current synchronous behaviour that we're relying on.

To avoid any issues:
- the timeout of the child process is set to `500ms`
- we check the output with `net.isIPv4` and carefully reject any abnormal results and process exits
- we check the socket's IP address against non-internal assignments in `os.networkInterfaces()`

**This method will currently only be used as a fallback method**

# Test Plan

Tested in isolation by running `getRouteIPAddress` with Node directly (this was tested both online and offline):
- Manually tested on Windows (Node.js 20, on Windows 11)
- Manually tested on macOS (Node.js 22, aarch64)
- Manually tested on Linux (Node.js 20 & 22, x86)

Tested with full `@expo/cli` by copying the new `build/src/utils/ip.js` build file into a local installation of `@expo/cli`, commenting out `internalIp.v4.sync()`, and running `expo start`.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/src/utils/ip.ts [hunk 1]
```diff
@@ -1,5 +1,81 @@
...
+
+/** Gets a route address by opening a UDP socket to a publicly routed address.
+ * @privateRemarks
+ * This is wrapped in `spawnSync` since the original `getIpAddress` utility exported
+ * in this module is used synchronosly. An appropriate timeout has been set and UDP
+ * ports don't send a message when opened.
+ */
+function getRouteAddress(): string | null {
+  const command = process.platform === 'win32' ? basename(process.execPath) : process.execPath;
+  const { error, status, stdout } = spawnSync(command, ['-'], {
```
Comment (byCedric) [inline packages/@expo/cli/src/utils/ip.ts:15]: Do we need to try-catch this?

File: packages/@expo/cli/src/utils/ip.ts [hunk 2]
```diff
@@ -1,5 +1,81 @@
...
+import { dirname, basename } from 'node:path';
+
+/** Gets a route address by opening a UDP socket to a publicly routed address.
+ * @privateRemarks
+ * This is wrapped in `spawnSync` since the original `getIpAddress` utility exported
+ * in this module is used synchronosly. An appropriate timeout has been set and UDP
+ * ports don't send a message when opened.
+ */
+function getRouteAddress(): string | null {
+  const command = process.platform === 'win32' ? basename(process.execPath) : process.execPath;
```
Comment (byCedric) [inline packages/@expo/cli/src/utils/ip.ts:14]: Wait, shouldn't the `process.execPath` be correctly adjusted on Windows already?

**Review by keith-kurak — APPROVED** [review]
I tested this on a Windows 11 PC that was missing `wmic` (seems to be a thing on new installs since Fall last year), and it worked!

Also tried imperfectly simulating LAN-but-no-WAN by connecting to an IoT device's wifi and it still resolved fine.

I've casually observed in the past that `npx expo start` is slower to start on PowerShell, so wanted to check if there was any performance impact of the change by timing from running the command to when a QR code appears. No difference with the added fallback:
- PS / no-fallback: 12 seconds
- PS / fallback: 12 seconds
- PS / fallback / no-internet-LAN: 13 seconds
- PS / fallback / no-internet-LAN / `--offline`: 4 seconds
(doesn't matter here, but curiosity got the best of me)
- WSL: 2 seconds
- macOS: 2 seconds
(maybe a mystery to solve for another day?)

File: packages/@expo/cli/src/utils/ip.ts (hunk 2 above)
Comment (kitten) [reply packages/@expo/cli/src/utils/ip.ts:15]: actually, maybe this is related to `windowsVerbatimArguments`. Disabling that caused issues with prior arguments I passed, but maybe that can be flipped again 🤔 I'll have a quick check

Comment (kitten) [reply packages/@expo/cli/src/utils/ip.ts:15]: ha, yea, that fixes it. I'll flip it back

**Review by EvanBacon — APPROVED** [review]

**Pull request merged**